src/utils/bounded_priority_queue.rs

```diff
@@ -0,0 +1,57 @@
+use crate::serializable_traits::Data;
+use serde_derive::{Deserialize, Serialize};
+use std::collections::BinaryHeap;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct BoundedPriorityQueue<T: Ord> {
+    max_size: usize,
+    underlying: BinaryHeap<T>,
+}
+
+impl<T: Data + Ord> Into<Vec<T>> for BoundedPriorityQueue<T> {
+    fn into(self) -> Vec<T> {
+        let mut col: Vec<_> = self
+            .underlying
+            .into_iter_sorted()
+            .collect();
+        col.reverse();
+        col
+    }
+}
+
+impl<T: Data + Ord> BoundedPriorityQueue<T> {
+    pub fn new(max_size: usize) -> BoundedPriorityQueue<T> {
+        BoundedPriorityQueue {
+            max_size: max_size,
+            underlying: BinaryHeap::with_capacity(max_size),
+        }
+    }
+
+    pub fn merge(&mut self, other: BoundedPriorityQueue<T>) -> &Self {
+        other
+            .underlying
+            .into_iter()
+            .for_each(|elem| self.append(elem));
+        self
+    }
+
+    pub fn append(&mut self, elem: T) {
+        if self.underlying.len() < self.max_size {
+            self.underlying.push(elem);
+        } else {
+            self.maybe_replace_lowest(elem);
+        }
+    }
+
+    pub(self) fn maybe_replace_lowest(&mut self, elem: T)
```

Comment:
`pub(self)` here would be equivalent to `fn` so is a bit redundant (and it doe snot require to be pub).

Reply:
Aha I was searching of how to define private methods and this came to my sight. Just forgot I could simply not declare `pub` haha. Will remove now.

Reply:
yes by default rust is "module private", which is a great default to have hehe